packages/payments-plugin/src/mollie/mollie.helpers.ts

```diff
@@ -0,0 +1,79 @@
+import { CreateParameters } from '@mollie/api-client/dist/types/src/binders/orders/parameters';
+import { Amount } from '@mollie/api-client/dist/types/src/data/global';
+import { OrderAddress as MollieOrderAddress } from '@mollie/api-client/dist/types/src/data/orders/data';
+import { Customer, Order } from '@vendure/core';
+
+import { OrderAddress } from './graphql/generated-shop-types';
+
+/**
+ * Check if given address has mandatory fields for Mollie Order and map to a MollieOrderAddress.
+ * Returns undefined if address doesn't have all mandatory fields
+ */
+export function toMollieAddress(address: OrderAddress, customer: Customer): MollieOrderAddress | undefined {
+    if (address.city && address.countryCode && address.streetLine1 && address.postalCode) {
+        return {
+            streetAndNumber: `${address.streetLine1} ${address.streetLine2 || ''}`,
+            postalCode: address.postalCode,
+            city: address.city,
+            country: address.countryCode.toUpperCase(),
+            givenName: customer.firstName,
+            familyName: customer.lastName,
+            email: customer.emailAddress,
+        };
+    }
+}
+
+/**
+ * Map all order and shipping lines to a single array of Mollie order lines
+ */
```

Comment:
@michaelbromley Will this always result in the order total? Mollie will validate if all lines add up to the order total, including taxes.

I don't like recreating this logic, but I've seen this a few times now: accounting software and external fulfillment platforms that want all order lines with prices, taxes and discounts AND want it to add up to the total you also supply.

Reply:
I have included shipping in the calculation, but surcharges...

Are surcharges included in the `line.proratedLinePrice` price? 

Mollie only knows orderLines, it doesn't know the concept of surcharges. The docs state this:

> The total amount of the line, including VAT and discounts. Adding all totalAmount values together should result in the same amount as the amount top level property.

So, the sum of `lines.totalAmount` must equal the `order.totalAmount`.

Reply:
> Are surcharges included in the line.proratedLinePrice price?

No, they are separate.

In that case then yes, map surcharges to lines and Mollie should be happy.

Reply:
@michaelbromley Not really familiar with surcharges. How can I add a promotion that will add a surcharge to the order? I would like to test if Mollie allows negative numbers on an order line + include a surcharge in the e2e test.

Reply:
Surcharges are not related to Promotions. Promotions can only make "adjustments", which is what we already take into account in the `discountedPrice` and `proratedPrice`. However, promotions _can_ make an orderline negative - this is an edge case and almost always an error in the PromotionAction logic. But should be gracefully handled if it comes up.

Surcharges must be manually added using `OrderService.addSurchargeToOrder`. There is currently no built-in mutation for this.

Reply:
Thanks, I will try to fix this today.